worker/uniter/paths.go

```diff
@@ -107,25 +112,39 @@ type StatePaths struct {
 
 // NewPaths returns the set of filesystem paths that the supplied unit should
 // use, given the supplied root juju data directory path.
-func NewPaths(dataDir string, unitTag names.UnitTag) Paths {
-	return NewWorkerPaths(dataDir, unitTag, "")
+func NewPaths(dataDir string, unitTag names.UnitTag, isRemote bool) Paths {
+	return NewWorkerPaths(dataDir, unitTag, "", isRemote)
 }
 
 // NewWorkerPaths returns the set of filesystem paths that the supplied unit worker should
 // use, given the supplied root juju data directory path and worker identifier.
 // Distinct worker identifiers ensure that runtime paths of different worker do not interfere.
-func NewWorkerPaths(dataDir string, unitTag names.UnitTag, worker string) Paths {
+func NewWorkerPaths(dataDir string, unitTag names.UnitTag, worker string, isRemote bool) Paths {
 	join := filepath.Join
 	baseDir := join(dataDir, "agents", unitTag.String())
 	stateDir := join(baseDir, "state")
 
-	socket := func(name string, abstract bool) string {
-		if os.HostOS() == os.Windows {
+	newSocket := func(name string, abstract bool) sockets.Socket {
+		podIP := os.Getenv(provider.OperatorPodIPEnvName)
+		if isRemote && podIP != "" {
+			switch name {
+			case "agent":
+				return sockets.Socket{
+					Network: "tcp",
+					Address: fmt.Sprintf("%s:%d", podIP, jujucServerSocketPort+unitTag.Number()),
```

Comment:
Future work, but we can't guarantee that a port is free. The logic should probably change to a random in range with retry mechanism, or incrementing port number with retry.

Reply:
so the idea is each unit has its own port assigned which means we do have a max number of units supported per application.
We can guarantee no port conflict because one unit can only have one action running at the same time, I think.
Rather, it will have a very high chance to get port conflict if we assign random port for units when we got a lot of units for an application.

Reply:
I mean, from within jujud we can guarantee that we don't share ports, but another process could bind to a port we are expecting to be unbound. Personally, I'm not a fan of binding a socket for every server. We can resolve this in the TLS work to share a single listener across many units.

Otherwise the options are to bind all ports at start-up (which is expensive, too many open files), or implement retry logic (which also has no guarantee we have any free ports).

Reply:
yeah, currently logic is one socket per unit.
Ideally, it would be good to change it to listen to 1 socket and include the unit tag as an arg in the request when we implement the auth.

Reply:
We will be changing actions to allow concurrent execution. We do need to cater for handling the port in use scenario.

Reply:
added a card https://trello.com/c/YXAqotWJ